Core candy machine

tonyboylehub opened the pull request:
PR for the first initial candymachine integration which covers the creation of a the candymachine via both `--wizard` and through individual steps.

There is also a quick template feature added to toolbox too which just git clones some known templates from our repo which I found handy.

## CHANGELOG.md

### Added
- **Candy Machine Commands**: New `cm` command group for creating and managing MPL Core Candy Machines
  - `mplx cm create` - Create candy machines with multiple modes:
    - `--wizard` - Interactive wizard for guided candy machine creation
    - `--template` - Generate candy machine template directory
    - Direct creation from existing configuration
  - `mplx cm upload` - Upload candy machine assets to decentralized storage
  - `mplx cm insert` - Insert items into existing candy machines
  - `mplx cm validate` - Validate asset cache and uploads
- **Template Commands**: New `toolbox template` commands for downloading project templates
  - `mplx toolbox template program` - Download Solana program templates (Shank)
  - `mplx toolbox template website` - Download website templates (Next.js with Tailwind/ShadCN)

### Enhanced
- **Interactive Wizard**: Comprehensive candy machine creation wizard with:
  - Asset discovery and validation
  - Collection configuration (create new or use existing)
  - Guard and group setup for minting rules
  - Progress tracking for uploads and transactions
  - Detailed completion summary
- **Asset Management**: Improved asset handling with:
  - Batch upload support for images and metadata
  - Asset cache management for efficient re-uploads
  - Validation of asset integrity and completeness
- **Configuration System**: Enhanced candy machine configuration with:
  - JSON-based configuration files (`cm-config.json`)
  - Guard configuration support (SOL payments, allowlists, etc.)
  - Group-based minting phases
  - Collection integration

### Technical Improvements
- **Transaction Handling**: Improved transaction building and confirmation for candy machine operations
- **Error Handling**: Enhanced error messages and validation throughout candy machine workflow
- **Testing**: Comprehensive test suite for candy machine functionality including:
  - Unit tests for candy machine creation
  - Integration tests for full lifecycle (create, upload, insert)
  - Test helpers for common operations

### Documentation
- **Candy Machine Guide**: Complete documentation covering:
  - Directory structure and asset organization
  - Wizard and manual workflow options
  - Guard and group configuration examples
  - Best practices and troubleshooting
- **Template System**: Quick-start templates for Solana development:
  - Program templates for building on-chain logic
  - Website templates for frontend development

### Dependencies
- Updated to `@metaplex-foundation/mpl-core-candy-machine` v0.3.0
- Enhanced UMI integration for candy machine operations
- Improved file upload handling with progress indicators

tonyboylehub commented on src/commands/toolbox/storage/upload.ts:
```diff
@@ -0,0 +1,88 @@
+            progressSpinner.succeed('Upload completed')
+
+
+            const uploadCache: UploadCache = {
+                files: uploadResult.map(result => ({
+                    fileName: result.fileName,
+                    uri: result.uri || '',
+                    mimeType: result.mimeType,
+                })),
+            }
```
this is the correct behaviour

MarkSackerberg commented:
This is awesome! I'm testing quite extensively currently and listing all the minor stuff that I've found which are mostly description nitpicks or minor other stuff.

Some general things I've found:
- It's a bit strange that when just running the mplx command the description of `cm` mentions two different methods with a colon, but then in the next line a different command is described
<img width="1004" height="598" alt="image" src="https://github.com/user-attachments/assets/04fcac5c-8f21-456b-a23f-e9019b6f9362" />
- `mplx cm --help` mentions two methods but four are described. Also shouldn't that be the description for `mplx cm create`?
<img width="1624" height="986" alt="image" src="https://github.com/user-attachments/assets/ec2e2c68-6cb2-4ee6-a735-2fa471d00a34" />
- Nit: The description in `mplx cm validate --help` is `Validate the asset cache file`. That doesn't really tell me what it's validating? Like is it checking that all config lines are written according to the cache file? Or something else?
- Two Nits For `mplx cm withdraw --help` 

- we should add a warning that this command makes the Candy Machine unusable 

-  `--address=<value>  The address to withdraw the candy machine to` Does this mean it is the destination of the rent? In that case we should clarify and maybe add another parameter to the roadmap to pass in the cm address (for the ones who do not have their config anymore)

- Nit: When running `mplx cm create --wizard` without assets in the asset folder and the user not adding assets it might be great to point out that there is a template command to create some defaults.
<img width="1842" height="658" alt="image" src="https://github.com/user-attachments/assets/5ee44f4d-e8e7-460d-b408-797b1c21c214" />

- Nit: in the wizard `✔ Should the NFTs be mutable? (y/n or q to quit)` we should probably recommend to have them mutable.
- Nit: in the wizard can we order the guards alphabetically? E.g. I was looking for assetBurn which is quite at the bottom, so I first thought it's not part of the wizard.
- Nit: For botTax we allow choosing `lastInstruction` true or false. We might want to add a warning that this could fail for some wallets since they are adding instructions at the end of the tx (lighthouse)?
<img width="1632" height="50" alt="image" src="https://github.com/user-attachments/assets/3b025d03-a1d4-45b8-928e-74d929658a66" />
- For date specific guards (startDate, endDate) we might want to suggest a date format and which timezone is used (UTC). e.g. I would have absolutely expected this to work:
<img width="2022" height="620" alt="image" src="https://github.com/user-attachments/assets/c9039398-af68-4372-b028-2922db9dcda8" />


- Nit: using the wizard the summary at the end has a line `Collection: Collection`. I don't think this line is of any use?
<img width="1020" height="212" alt="image" src="https://github.com/user-attachments/assets/ada6910c-0783-4d47-a98b-6a79c9ef6a21" />

- Potential Bug: In the wizard when no guard is added (neither global nor group) there is a warning, which is good. But no cm-config.json is created? I assume we would need that for withdrawing or adding guards though.

tonyboylehub commented:
- fixed help descriptions.
- Added template description when trying to run --wizard without any assets in folder.
- added clarifications for date format.
- nfts recommended as mutable\
- guard list alphabetize
- candymachine withdraw arg description fixed. 

- Summary at end of the wizard mentioned `Collection: Collection`, this is actually the collection name from the collection.json file. `Collection: <collection.json.name>`. It just so happens that in my tests the collection is named "Collection".

MarkSackerberg commented:
when using the wizard the comment for mutability is now:
> [Recommend: y for flexibility]
shouldnt it be
> [Recommended: y for flexibility]
?

tonyboylehub commented:
@MarkSackerberg 

> The Date format that is proposed is not working but throwing a JS error:
Should be fixed.

>What should be the folder the CLI uses as base? With sugar the current folder was used. this seems not to be the case for cli:

Wizard now works based off current directory with a slightly different set up flow. you'll need to use `node ../bin/run.js cm create --wizard` after cd'ing into a test directory.

> How would I wrap a CM with a candy guard that has been created with "mplx cm create"? Are we missing a "mplx cm wrap" or similar with the manual steps?

This was a great catch. I was creating the CM's with empty candy guards if no guards where present. Now if no guards are present during wizard creation (or normal creation) a candy guard will not be created allowing on the auth to mint. If guards are present globally or in a group then a candy guard should be spun up and wrapped.

> How would I wrap a CM with a candy guard that has been created with "mplx cm create"? Are we missing a "mplx cm wrap" or similar with the manual steps?

updating/wrapping/unwrapping will have to come next as a priority due the guard and minting situation for security.

MarkSackerberg commented on src/lib/cm/prompts/createCandyMachineWizardPrompt.ts:
```diff
@@ -275,7 +283,7 @@ const createCandyMachinePrompt = async (): Promise<{ candyMachineConfig: CandyMa
             const groupGuards: RawGuardConfig = {}
 
             const selectedGuards: string[] = await checkbox({
-                message: `Select the guards to assign to group ${groupName} (press space to select, or select Quit to abort):`,
```
Why did you remove "(press space to select, or select Quit to abort)" for both occurrences? I personally preferred to have it in for ease of use.
